### ISSUE TITLE
Improve visuals of focus

### ISSUE THREAD

**Issue by bluenote10** [khonsulabs/cushy#203]
While the focus handling works very well for me in terms of functionality, its visuals are a bit odd looking.

Let's take the example of a checkbox, which before clicking / taking focus looks like:

![image](https://github.com/user-attachments/assets/e56957f8-2ac7-4272-b385-ee726b572fdf)

So far it has a clean / lightweight look to it. During hover itself, it keeps its modern look:

![image](https://github.com/user-attachments/assets/9dbe45f8-e7ba-460e-8bc6-a455a2d84d9b)

However, immediately when clicking it, it takes focus which changes its visual to something rather "heavy", giving it a somewhat inconsistent look:

![image](https://github.com/user-attachments/assets/d56717c6-fcfc-449d-ac2b-9ac2f477f784)

Here are some ideas how to improve the visuals.

### Improve via theme

Partly the issue is caused by the rather think lines, which seem to be covering 4 pixels in this case:

![image](https://github.com/user-attachments/assets/944c45e7-281e-447f-a765-ff6b8751a4ad)

I haven't checked yet whether the current theming system would allow changing them. However, I'd assume that a designer would generally complain about the suddenly introduced border itself, possible for a few reasons:

- It puts an awkward emphasis on the trailing whitespace in the label (which typically are needed to align multiple widgets in a form).
- It makes the label less readable (due to the border conflicting with font descenders).
- The non-rounded border doesn't play nicely with the rounded button outline?

### Allow turning them off consistently

It largely depends on the use case whether controlling the UI via the keyboard is essential or not. In my use cases it would be a rarely used feature, which is why I was considering to turn off focus handling in the hope to improve the visuals.

While looking at how to solve that in this particular example, I noticed that `Button` has `prevent_focus()` as a regular method, but a `Checkbox` actually has not. Looking at how `Checkbox` constructs its underlying button:

https://github.com/khonsulabs/cushy/blob/d032299fc584b64bbb07daef1d17f14ff154125b/src/widgets/checkbox.rs#L56-L67

it would be possible to forward the "prevent focus" information into `Checkbox::new()` or `.into_checkbox()`, but that would not be very idiomatic (because no builder pattern).

A more idiomatic solution would be to offer `prevent_focus()` on `Checkbox` itself, which forwards the call to its underlying button (which would probably require holding a reference to the underlying button). This made me wonder if `prevent_focus()` should actually be a trait method on `Widget` to have a generic solution for "disabling focus handling for this entire widget"?

### Smart focus handling

This made me wonder why focus highlighting is less disturbing in other GUIs, and started looking at prior art.

I noticed that many GUIs / websites seem to render focus highlights only "on demand", which feels pretty neat. The idea is basically that the focus highlights are mostly hidden, _even after clicking a button_. For instance when I click the "Write" and "Preview" buttons in the GitHub text form, they don't get highlight rectangles:

![image](https://github.com/user-attachments/assets/7ab25e1b-8df7-484d-9173-7d1938816dfc)

Only after pressing `<TAB>` once the currently focused element gets revealed via the visual hint:

![image](https://github.com/user-attachments/assets/8e1177a7-55c6-416c-8f60-a89519045346)

From there the highlights are rendered until a blur occurs, e.g. the user clicks into no man's land. 

This feels pretty natural, because when switching from using the mouse for controlling the UI to the keyboard, a single `<TAB>` stroke is needed. Typically when navigating with the keyboard, one is used to hitting the `<TAB>` key frequently, so the overhead of the initial key press is negligible. At the same time, when navigating a UI purely with the mouse, one simply doesn't care about what has the focus (because it basically has no effect in pure mouse navigation). This allows to keep the "distracting" focus hints hidden by default, and only opt-in into rendering them when actually needed.

I would assume that such a smart focus handling would require a bit more work though.

**Comment by ecton** [comment]
> I haven't checked yet whether the current theming system would allow changing them. 

Currently the width is hard-coded, but it should be a component that can be customized. The color can be customized.

>    It puts an awkward emphasis on the trailing whitespace in the label (which typically are needed to align multiple widgets in a form).

This can be solved by aligning the button to the left rather than having it fill its parent.

>    It makes the label less readable (due to the border conflicting with font descenders).

This is very valid feedback. The focus ring shouldn't occlude the focused content.

>    The non-rounded border doesn't play nicely with the rounded button outline?

This is also a good point. The way the Checkbox is currently implemented is trying to reuse a lot of the Button logic, but it means that it's created as a transparent button. I like how this makes things very consistent, but the focus ring is a challenge because it would be *nice* to render the focus ring around the checkbox itself. It might be that the Checkbox should really just become its own widget.

But in the meantime, I think switching it to honor the button corner radius and spacing it a little bit away from the button content will solve most of the issues.

> Allow turning them off consistently

`somewidget.with(&HighlightColor, Color::CLEAR_WHITE)` will cause `somewidget` and all of its children to not render keyboard focus indicators while still allowing keyboard accessibility. Keyboard accessibility is incredibly important and should almost never be disabled.

> For instance when I click the "Write" and "Preview" buttons in the GitHub text form, they don't get highlight rectangles:

That's not what's happening. When clicking on Write, focus is advanced to the tab content. Github has tied the next focusable control after the compose widget to be the tab tabs, which is why hitting tab advances focus to the next widget. 

With that being said, I agree checkboxes look a bit heavy handed when focused. I disagree with the premise that the widget that has keyboard focus should *not* show a focus indicator of some sort. To me, if the user can't determine if hitting the space bar will trigger an interaction, the user interface has failed the user.

Checkboxes take keyboard focus, even in web browsers, but they have inconsistent focus-showing rules. However, on all Once clicked, hitting the space bar will continue to toggle the checkbox and hitting tab will advance focus to the next focusable widget. However, the fact that browsers all have different behaviors of how focus is drawn really shows why trying to compare behavior to web browsers is not actually the best thing to do.

My current plan of action is:

- Make the focus ring width be customizable via the style system
- Transparent button focus should honor the button corner radius's and have some padding between the content and the ring
- Consider making the focus ring be partially transparent rather than solid to make it less harsh.

**Comment by ecton** [comment]
> My current plan of action is:
> 
> * Make the focus ring width be customizable via the style system
> * Transparent button focus should honor the button corner radius's and have some padding between the content and the ring
> * Consider making the focus ring be partially transparent rather than solid to make it less harsh.

I've implemented these changes, and I think it's a drastic improvement:

Before:
![Before](https://github.com/user-attachments/assets/161637bc-8a3c-4923-99d0-3658a3409731)

After:
![After](https://github.com/user-attachments/assets/6a67a4bd-a0b9-46b1-9727-489ee46fa260)

Let me know what you think!

**Comment by bluenote10** [comment]
> Let me know what you think!

That certainly is a big improvement!

> I disagree with the premise that the widget that has keyboard focus should not show a focus indicator of some sort. To me, if the user can't determine if hitting the space bar will trigger an interaction, the user interface has failed the user.

Note that the commonly implemented behavior is somewhat different:
- If there is no visual focus, hitting `<spacebar>` has no effect.
- Only if the user has entered visual focus mode via `<tab>`, hitting `<spacebar>` does something.

I.e., it should not be the case that pressing `<spacebar>` triggers some unknown (because not visually highlighted) element.

As far as I can see there has been a very clear trend towards using explicit focus mode semantics. I personally prefer this mode a lot, because of the clear separation between "I'm in _mouse mode_" (no focus visible) vs "I'm in _keyboard mode_" (focus visible). Mentally, pressing `<tab>` enters the _keyboard mode_, which is super natural. Conversely, clicking somewhere switches back to _mouse mode_, because clearly the user has started using the mouse again. I don't expect `<spacebar>` to do anything in _mouse mode_, because it is visually obvious (no focus highlighted) that I'm not in _keyboard mode_.

Note that the explicit mode also enables using `<spacebar>` for keyboard shortcut purposes. Think for instance about media players that use the `<spacebar>` as the standard mechanism for play/pause. A typical UX flaw that I used to encounter in the past with UIs that implicitly give focus: I'm watching a video, and at some point I need to quickly mute and unmute the video for whatever reason using the mouse. Now the UI has implicitly transferred the focus to the mute button as a result of this. Minutes later I want to pause the video by pressing `<spacebar>`. But that is not what will happen: Instead I will mute the video again, even though it was never my intention to focus that button. An explicit keyboard focus mode allows for a better UX here, because it doesn't "accidentally" change the semantics of `<spacebar>`.

-----

For what it's worth, I'm collecting a few data points on apps and websites (as far as can tell the focus/highlight handling on a website is not determined by the browser, but rather by CSS -- which may or may not be browser specific). In all these examples I check whether clicking a checkbox/toggle automatically gives it a visual focus, and whether using `<tab>` is the standard mechanism to enter keyboard mode.

### **Google Chrome**

Checking its native behavior in the settings dialog:

![image](https://github.com/user-attachments/assets/aeccc171-8c48-4b8d-a830-a62fbbdee4f7)

Clicking checkbox/toggle gives it focus: **No**
Pressing `<tab>` enters keyboard focus mode: **Yes**

### **Firefox**

Checking its native behavior in the settings dialog:

![image](https://github.com/user-attachments/assets/534c515b-3b21-4e39-a23a-e6a93d135fab)

Clicking checkbox/toggle gives it focus: **No**
Pressing `<tab>` enters keyboard focus mode: **Yes**

### **VSCode**

![image](https://github.com/user-attachments/assets/cc7c2eb6-8b3c-403b-8a15-72c422332c0e)

Clicking checkbox/toggle gives it focus: **Yes**

VSCode seems to be a notable exception where clicking always takes focus. Note however that its focus highlight is very subtle, because it doesn't add now borders, but rather just changes the color of the existing borders.

### **GTK applications**

I'm using `nautilus` as an example, but I assume that this applies for basically all Gnome apps based on GTK (and thus all standard applications in Ubuntu).

![image](https://github.com/user-attachments/assets/a0b4249c-cbb7-4ddf-ad26-a6752c9364ba)

Clicking checkbox/toggle gives it focus: **No**, without having entered keyboard mode before, but **Yes** after having been in keyboard mode in that particular app.
Pressing `<tab>` enters keyboard focus mode: **Yes**

The same seems to apply for toggles in GTK.

### **Material Design (on example GMail)**

In the settings and also for the checkboxes in the inbox:

![image](https://github.com/user-attachments/assets/9b31dc54-ee62-4c4f-92bc-d4981fb57007)

Clicking checkbox/toggle gives it focus: **No**
Pressing `<tab>` enters keyboard focus mode: **Yes**

However, in its search/filter form it does it differently:

![image](https://github.com/user-attachments/assets/902c869d-e9e1-4135-9cd7-cb2549f08025)

Clicking checkbox/toggle gives it focus: **Yes**

### **GitHub**

E.g. in the notifications:

![image](https://github.com/user-attachments/assets/75e68890-8ad4-479d-b14e-0d27758d977e)

Clicking checkbox/toggle gives it focus: **No**
Pressing `<tab>` enters keyboard focus mode: **Yes**

### **Steam**

![image](https://github.com/user-attachments/assets/9aab3476-7306-43d9-95ed-b0816d406636)

Clicking checkbox/toggle gives it focus: **No**
Pressing `<tab>` enters keyboard focus mode: **No**, Steam doesn't seem to have a keyboard selection mode as far as I can tell.

---------------

The picture I got from my small investigation is that there is certainly no clear standard when it comes to focus handling, and that some websites even mix it up internally. Nonetheless, I have the impression that the "enter explict keyboard docus mode via `<tab>`" pattern has become more prominent over the past few years. I personally feel that the separation of mouse vs keyboard modes leads to an improved UX.

**Comment by ecton** [comment]
All of these examples are kind of moot -- every application can be coded to do whatever they want. Additionally, they may behave differently depending on what platform you're on. For example, you've said that Github behaves one way, but here's my video evidence showing that the checkboxes *do receive focus* on Windows in Firefox:

![Recording 2024-11-02 133748](https://github.com/user-attachments/assets/5dc757a1-965b-46cc-be06-970b87e8049d)

I first clicked on the checkbox, then used the space bar to toggle the checkbox. This shows that keyboard focus was indeed set to the checkbox even though no visual indicator is shown. This is how Chrome behaves on Windows as well on the same notifications list.

Keyboard accessibility is *very important to me*, and I want all Cushy apps to have [full accessibility](#118) eventually. This hybrid approach that's being proposed sounds nice at first glance, but it adds potential confusion for an end user all in the sake of making an app a little prettier for people who don't ever use keyboard-interactions for certain types of widgets. Why does a text field always show a focus indicator when it is the focused widget, yet other controls that respond to keyboard interactions don't until the tab key is pressed? It seems arbitrary and a step in the wrong direction for legible and predictable user interfaces.

**Comment by ecton** [comment]
> Allow turning them off consistently

While refactoring the Checkbox I did discover I had already thought about disallowing focus but forgot one of the features I had added. On Mac OS, there's a system setting that limits or enables keyboard accessibility. By default, tabbing through controls on Mac OS doesn't focus regular push buttons or checkboxes or radios. In preparation for supporting that setting, there is a component `AutoFocusableControls`. Your original request of disabling them consistently is actually doable today by using `.with(&AutoFocusableControls, FocusableWidgets::OnlyTextual)`.

That being said, I'm adding a final mode to the checkbox widget that allows it to use its own standalone checkbox indicator rather than piggybacking the button. This will allow the focus indicator to be drawn stylistically on the checkbox itself only, which I think should alleviate your major issues with the current implementation while still supporting what I think is the correct approach to keyboard accessibility.

**Comment by ecton** [comment]
Here's the result of my changes to refactor radio buttons and checkboxes into their own widgets. I did this by adding a new widget type `Indicator` that should be able to be used to add the on-off toggles that some UIs are using these days.

Unchecked, unfocused:
![image](https://github.com/user-attachments/assets/06b474ca-5591-48d3-ae22-52075022310b)

Unchecked, focused:
![image](https://github.com/user-attachments/assets/dd59f3d2-cf92-4cca-971d-4285a62b931a)

Checked, focused:
![image](https://github.com/user-attachments/assets/1b6fab0c-81bc-48fa-b1a5-43e97cf7b373)

Checked, unfocused:
![image](https://github.com/user-attachments/assets/6bdcc2ea-ee80-4b07-b76b-491b66befcfd)

I think this set of changes should alleviate most concerns about focus indicators on these widget types, but when paired with the other features Cushy has now, I think this issue can be considered resolved with maybe a feature request to implement this other delayed-focus-style that some platforms seem to be adopting.

**Comment by ecton** [comment]
I split the remaining feature request into its own issue for consideration and others to weigh in on as well. Thank you again for your feedback and helping improve Cushy!

**Comment by bluenote10** [comment]
> I think this issue can be considered resolved with maybe a feature request to implement this other delayed-focus-style that some platforms seem to be adopting.

Yes, absolutely. The improved checkbox rendering looks really great, thanks for the enhancements!
